// ngx-typeahead/esm5/lib/ngx-typeahead.utils.js
import { __assign } from "tslib";
import { HttpParams } from '@angular/common/http';
import { Key } from './models';
export function validateNonCharKeyCode(keyCode) {
    return [
        Key.Enter,
        Key.Tab,
        Key.ShiftLeft,
        Key.ShiftRight,
        Key.ArrowLeft,
        Key.ArrowUp,
        Key.ArrowRight,
        Key.ArrowDown,
        Key.MetaLeft,
        Key.MetaRight,
    ].every(function (codeKey) { return codeKey !== keyCode; });
}
export function validateArrowKeys(keyCode) {
    return keyCode === Key.ArrowDown || keyCode === Key.ArrowUp;
}
export function isIndexActive(index, currentIndex) {
    return index === currentIndex;
}
export function isEnterKey(event) {
    return event.code === Key.Enter;
}
export function isEscapeKey(event) {
    // tslint:disable-next-line: deprecation
    return event.code === Key.Escape;
}
export function createParamsForQuery(query, queryParamKey, customParams) {
    var _a;
    if (queryParamKey === void 0) { queryParamKey = 'q'; }
    if (customParams === void 0) { customParams = {}; }
    var searchParams = __assign((_a = {}, _a[queryParamKey] = query, _a), customParams);
    // tslint:disable-next-line
    var setParam = function (acc, param) {
        return acc.set(param, searchParams[param]);
    };
    var params = Object.keys(searchParams).reduce(setParam, new HttpParams());
    return params;
}
export function resolveApiMethod(method) {
    if (method === void 0) { method = ''; }
    var isMethodValid = [
        'get',
        'post',
        'put',
        'delete',
        'patch',
        'request',
    ].some(function (methodName) { return method === methodName; });
    var apiMethod = isMethodValid ? method : 'get';
    return apiMethod;
}
export var NO_INDEX = -1;
export function resolveNextIndex(currentIndex, stepUp, listLength) {
    if (listLength === void 0) { listLength = 10; }
    var step = stepUp ? 1 : -1;
    var topLimit = listLength - 1;
    var bottomLimit = NO_INDEX;
    var currentResultIndex = currentIndex + step;
    var resultIndex = currentResultIndex;
    if (currentResultIndex === topLimit + 1) {
        resultIndex = bottomLimit;
    }
    if (currentResultIndex === bottomLimit - 1) {
        resultIndex = topLimit;
    }
    return resultIndex;
}
export function toJsonpSingleResult(response) {
    return response[1];
}
export function toJsonpFinalResults(results) {
    return results.map(function (result) { return result[0]; });
}
export function hasCharacters(query) {
    return query.length > 0;
}
export function toFormControlValue(e) {
    return e.target.value;
}
export function resolveItemValue(item, fieldsToExtract, caseSensitive) {
    if (caseSensitive === void 0) { caseSensitive = false; }
    var newItem = item;
    if (!item.hasOwnProperty('length')) {
        var fields = !fieldsToExtract.length
            ? Object.keys(item)
            : fieldsToExtract;
        newItem = fields.reduce(function (acc, cur) { return "" + acc + item[cur]; }, '');
    }
    return caseSensitive ? newItem : newItem.toLowerCase();
}
